### PR TITLE
[direct call] changes raylet to push tasks to worker

### PR DESCRIPTION
**Pull request by zhijunfu**
## What do these changes do?

The previous change adds a grpc server to worker, this pr changes raylet to assign tasks instead of waiting for workers to get tasks.

## Related issue number

#5029 

## Linter

- [Y] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by zhijunfu** [comment]
@raulchen @jiangzihao2009

File: src/ray/core_worker/task_execution.cc [hunk 1]
```diff
@@ -7,64 +7,95 @@ namespace ray {
...
+
+  auto status = execution_callback_(func, args, task_info, num_returns);
+  // TODO(zhijunfu):
+  // 1. Check and handle failure.
+  // 2. Save or load checkpoint.
+  return status;
+}
+
+Status CoreWorkerTaskExecutionInterface::Run() {
+  auto callback = [this](const raylet::TaskSpecification &spec) {
```
Comment (raulchen) [inline src/ray/core_worker/task_execution.cc]: This should be removed?

File: src/ray/rpc/server_call.h [hunk 2]
```diff
@@ -129,13 +128,9 @@ class ServerCallImpl : public ServerCall {
...
-    (service_handler_.*handle_request_function_)(request_, &reply_,
-                                                 [this](Status status) {
-                                                   // When the handler is done with the
-                                                   // request, tell gRPC to finish this
-                                                   // request.
-                                                   Finish(status);
-                                                 });
+    auto status = (service_handler_.*handle_request_function_)(request_, &reply_);
+    // When the handler is done with the request, tell gRPC to finish this request.
+    Finish(status);
```
Comment (raulchen) [inline src/ray/rpc/server_call.h]: The `done_callback` is used to allow the request handler to handle a request asynchronously. 
I think it's still needed, e.g., by `WaitRequst` and `FetchOrReconstructObjects`.

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -844,13 +844,36 @@ void NodeManager::ProcessClientMessage(
...
+      auto entry = worker_clients_.find(client_id);
+      if (entry != worker_clients_.end()) {
+        RAY_LOG(WARNING) << "Received registration of a new worker that already exists: "
+                         << client_id;
+        return;
+      }
+
+      // Initialize a rpc client to the new worker.
+      std::unique_ptr<rpc::WorkerTaskClient> grpc_client(
+          new rpc::WorkerTaskClient("0.0.0.0", worker->Port(), client_call_manager_));
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: The ip should be `127.0.0.1`?

File: src/ray/raylet/node_manager.cc [hunk 4]
```diff
@@ -1765,79 +1781,131 @@ bool NodeManager::AssignTask(const Task &task) {
...
+
+    // NOTE: we cannot directly call `finish_assign_task_callback` here because
+    // it assumes the task is in SWAP queue, thus we need to delay invoking this
+    // function after the assigned tasks are moved from READY queue to SWAP queue
+    // in `DispatchTasks`.
+    // Another option is to move the tasks to SWAP queue here just before calling
+    // `finish_assign_task_callback` so we can save an io_service post, at the
+    // expense of calling `MoveTask` for each of the assigned tasks.
+    io_service_.post(
+        [status, finish_assign_task_callback]() { finish_assign_task_callback(status); });
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: Does `status == ok` ensures that the request is successfully sent to the worker? 
I guess perhaps not, because `AssignTask` is async.

File: src/ray/raylet/node_manager.h [hunk 5]
```diff
@@ -454,10 +455,18 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   void HandleDisconnectedActor(const ActorID &actor_id, bool was_local,
                                bool intentional_disconnect);
 
+  /// Handle the case where a worker is available. This should be called when a new
+  /// worker is registered with a port greater than zero (which indicates it will
+  /// act as a rpc server), and when a worker finishes it assigned task, which is
+  /// indicated by a `GetTask` message (or a rpc reply if the worker is a rpc server).
+  ///
+  /// \param worker Worker that is available.
+  /// \return void.
```
Comment (raulchen) [inline src/ray/raylet/node_manager.h]: we should document clearly that pull mode is temporarily used for old workers, and eventually we'll only have push mode.

File: src/ray/raylet/node_manager.cc [hunk 6]
```diff
@@ -1765,79 +1781,131 @@ bool NodeManager::AssignTask(const Task &task) {
...
+    auto &client = client_entry->second;
+
+    rpc::AssignTaskRequest request;
+    request.set_task_id(task_id.Binary());
+    request.set_task_spec(task.Serialize());
+    request.set_resource_ids(resource_id_set.Serialize());
+
+    auto status = client->AssignTask(
+        request, [this, worker](Status status, const rpc::AssignTaskReply &reply) {
+          // Worker has finished this task. The logic below is the same
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: The task could run arbitrarily long. Will gRPC treat this as a request timeout?

File: src/ray/raylet/node_manager.h (hunk 5 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.h]: makes sense. thanks.

File: src/ray/rpc/server_call.h (hunk 2 above)
Comment (zhijunfu) [reply src/ray/rpc/server_call.h]: yes... I forgot that one.

File: src/ray/raylet/node_manager.cc (hunk 4 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: I think if server is already dead, then the client should be able to tell from the status. I can double check this. On the other side, server still can crash before it has any chance to process the request.

**Review by zhijunfu — COMMENTED** [review]
Thanks. Updated accordingly.

File: src/ray/raylet/node_manager.cc (hunk 3 above)
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: On the server side, we can use `0.0.0.0`, which means the server accepts connection from all IPs. But on the client side, this should be the server IP to connect. I don't understand why this can be `0.0.0.0`. How would the client find the server?

Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: while` 0.0.0.0` can work in the test, it's a kind of obscure in the materials I looked at. I will update it to `127.0.0.1` which is more clear in the meaning.

File: src/ray/raylet/node_manager.cc (hunk 4 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: Verified that the status can tell server failure if server crashes. I feel it's good enough in this case?

Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: I think as long as we can make sure the task won't be lost, it should be fine.

File: src/ray/core_worker/task_execution.cc [hunk 7]
```diff
@@ -7,65 +7,63 @@ namespace ray {
...
   }
 
+  auto status = execution_callback_(func, args, task_info, num_returns);
+  // TODO(zhijunfu):
+  // 1. Check and handle failure.
+  // 2. Save or load checkpoint.
+  return status;
+}
+
+Status CoreWorkerTaskExecutionInterface::Run() {
```
Comment (raulchen) [inline src/ray/core_worker/task_execution.cc]: Is the return status still needed?
And how can we propagate errors to Java now?

File: src/ray/raylet/node_manager.cc [hunk 8]
```diff
@@ -844,13 +844,36 @@ void NodeManager::ProcessClientMessage(
...
+
     worker_pool_.RegisterWorker(std::move(worker));
-    DispatchTasks(local_queues_.GetReadyTasksWithResources());
+    if (use_push_task) {
+      auto registered_worker = worker_pool_.GetRegisteredWorker(client);
+      HandleWorkerAvailable(registered_worker);
+    } else {
+      // TODO(zhijunfu): this seems unnecessary as we already call
+      // `DispatchTasks` when this worker calls `GetTask`?
+      DispatchTasks(local_queues_.GetReadyTasksWithResources());
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: I think it's unnecessary.

File: src/ray/raylet/node_manager.cc [hunk 9]
```diff
@@ -1765,79 +1782,131 @@ bool NodeManager::AssignTask(const Task &task) {
     worker->SetTaskResourceIds(acquired_resources);
   }
 
+  auto task_id = spec.TaskId();
+  auto finish_assign_task_callback = [this, worker, task_id](Status status) {
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: Code in this callback didn't change, right?

File: src/ray/raylet/node_manager.cc [hunk 10]
```diff
@@ -1765,79 +1782,131 @@ bool NodeManager::AssignTask(const Task &task) {
...
-              actor_entry->second.AddHandle(new_handle_id, execution_dependency);
-            }
-
-            // TODO(swang): For actors with multiple actor handles, to
-            // guarantee that tasks are replayed in the same order after a
-            // failure, we must update the task's execution dependency to be
-            // the actor's current execution dependency.
-          } else {
-            RAY_CHECK(spec.NewActorHandles().empty());
+  if (worker->Port() > 0) {
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: better wrap this in a method, say `worker->UsePush()`?

File: src/ray/raylet/node_manager.cc (hunk 4 above)
Comment (raulchen) [reply src/ray/raylet/node_manager.cc]: BTW, we can add a TODO here saying that when we fully migrate to push mode, this `post` and the `swap` queue won't be needed.

File: src/ray/raylet/node_manager.h [hunk 11]
```diff
@@ -454,6 +455,17 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   void HandleDisconnectedActor(const ActorID &actor_id, bool was_local,
                                bool intentional_disconnect);
 
+  /// Handle the case where a worker is available. This should be called when a new
```
Comment (raulchen) [inline src/ray/raylet/node_manager.h]: Nit, can you rephrase the comment like this?
```
This should be called in the following cases:
1) when ...
2) when ...
```

There are too many which-s and when-s in one sentence. I'm confused ..

File: src/ray/raylet/node_manager.cc [hunk 12]
```diff
@@ -1765,79 +1782,131 @@ bool NodeManager::AssignTask(const Task &task) {
...
+
+    rpc::AssignTaskRequest request;
+    request.set_task_id(task_id.Binary());
+    request.set_task_spec(task.Serialize());
+    request.set_resource_ids(resource_id_set.Serialize());
+
+    auto status = client->AssignTask(
+        request, [this, worker](Status status, const rpc::AssignTaskReply &reply) {
+          // Worker has finished this task. The logic below is the same
+          // as ProcessGetTaskMessage().
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Why not call `ProcessGetTaskMessage()` directly? It would be better if we can share as much code as possible with the `GetTask` codepath.

File: src/ray/raylet/node_manager.cc (hunk 10 above)
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc]: Can we actually move this whole if-else block to communicate with the worker into a separate method on the `Worker` class?

File: src/ray/raylet/node_manager.cc [hunk 13]
```diff
@@ -1765,79 +1782,131 @@ bool NodeManager::AssignTask(const Task &task) {
...
+    auto &client = client_entry->second;
+
+    rpc::AssignTaskRequest request;
+    request.set_task_id(task_id.Binary());
+    request.set_task_spec(task.Serialize());
+    request.set_resource_ids(resource_id_set.Serialize());
+
+    auto status = client->AssignTask(
+        request, [this, worker](Status status, const rpc::AssignTaskReply &reply) {
+          // Worker has finished this task. The logic below is the same
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Do we have any guarantees about when this callback would run in relation to callbacks for the same worker on the `io_service_`? For example, if the worker sends a `NotifyUnblocked` message that it is no longer blocked in a `ray.get` on the normal raylet socket, then completes an assigned task, can we guarantee that we get the former message from asio before the gRPC callback for the latter event?

If not, then we will need to figure out which handlers depend on this order guarantee. For instance, I think that the `NotifyUnblocked` handler assumes that there is an executing task that was blocked in a `ray.get`.

File: src/ray/raylet/node_manager.cc [hunk 14]
```diff
@@ -844,13 +844,36 @@ void NodeManager::ProcessClientMessage(
...
+      std::unique_ptr<rpc::WorkerTaskClient> grpc_client(
+          new rpc::WorkerTaskClient("127.0.0.1", worker->Port(), client_call_manager_));
+      worker_clients_.emplace(client_id, std::move(grpc_client));
+    }
+
     worker_pool_.RegisterWorker(std::move(worker));
-    DispatchTasks(local_queues_.GetReadyTasksWithResources());
+    if (use_push_task) {
+      auto registered_worker = worker_pool_.GetRegisteredWorker(client);
+      HandleWorkerAvailable(registered_worker);
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Can you leave a comment explaining why this is only necessary for the `use_push_task` case?

File: src/ray/raylet/node_manager.h [hunk 15]
```diff
@@ -518,12 +530,16 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
...
+  /// The `ClientCallManager` object that is shared by all `NodeManagerClient`s
+  /// as well as all `WorkerTaskClient`s.
   rpc::ClientCallManager client_call_manager_;
 
   /// Map from node ids to clients of the remote node managers.
   std::unordered_map<ClientID, std::unique_ptr<rpc::NodeManagerClient>>
       remote_node_manager_clients_;
+
+  /// Map from worker ids to rpc clients of workers.
+  std::unordered_map<ClientID, std::unique_ptr<rpc::WorkerTaskClient>> worker_clients_;
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.h]: This should go in `WorkerPool`, not `NodeManager`.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]

File: src/ray/core_worker/task_execution.cc (hunk 7 above)
Comment (zhijunfu) [reply src/ray/core_worker/task_execution.cc]: the status is not needed and can be removed. 
we're not propagating errors to java here, previously we just check the return value of the java/python callback, and return directly if there's an error.

File: src/ray/raylet/node_manager.cc (hunk 8 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: thanks.

File: src/ray/raylet/node_manager.cc (hunk 9 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: sure, that should look simpler. thanks.

File: src/ray/raylet/node_manager.h (hunk 11 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.h]: will update. thanks.

File: src/ray/core_worker/mock_worker.cc [hunk 16]
```diff
@@ -40,17 +36,23 @@ class MockWorker {
       // Write the merged content to each of return ids.
       for (int i = 0; i < num_returns; i++) {
         ObjectID id = ObjectID::ForTaskReturn(task_info.task_id, i + 1);
-        RAY_CHECK_OK(worker_.Objects().Put(return_value, id));
+        RAY_CHECK_OK(worker_->Objects().Put(return_value, id));
       }
       return Status::OK();
     };
 
+    worker_ = std::unique_ptr<CoreWorker>(
```
[the remark this answers is not in the document]
Comment (zhijunfu) [reply src/ray/core_worker/mock_worker.cc]: sounds good. thanks.

File: src/ray/core_worker/core_worker.h [hunk 17]
```diff
@@ -25,7 +25,8 @@ class CoreWorker {
   /// NOTE(zhijunfu): the constructor would throw if a failure happens.
   CoreWorker(const WorkerType worker_type, const ::Language language,
              const std::string &store_socket, const std::string &raylet_socket,
-             const JobID &job_id = JobID::Nil());
+             const JobID &job_id,
+             CoreWorkerTaskExecutionInterface::TaskExecutor execution_callback = nullptr);
```
[the remark this answers is not in the document]
Comment (zhijunfu) [reply src/ray/core_worker/core_worker.h]: yes that's for drivers. Will remove the default value.

File: src/ray/core_worker/task_execution.cc [hunk 18]
```diff
@@ -7,65 +7,63 @@ namespace ray {
...
+  // Run main IO service.
+  main_service_.run();
+
   // should never reach here.
   return Status::OK();
```
[the remark this answers is not in the document]
Comment (zhijunfu) [reply src/ray/core_worker/task_execution.cc]: sounds good.

File: src/ray/raylet/node_manager.cc (hunk 14 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: done

File: src/ray/raylet/node_manager.cc [hunk 19]
```diff
@@ -844,13 +844,36 @@ void NodeManager::ProcessClientMessage(
...
+    if (use_push_task) {
+      auto entry = worker_clients_.find(client_id);
+      if (entry != worker_clients_.end()) {
+        RAY_LOG(WARNING) << "Received registration of a new worker that already exists: "
+                         << client_id;
+        return;
+      }
+
+      // Initialize a rpc client to the new worker.
+      std::unique_ptr<rpc::WorkerTaskClient> grpc_client(
```
[the remark this answers is not in the document]
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: done.

File: src/ray/raylet/node_manager.h (hunk 15 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.h]: ok, updated.

File: src/ray/raylet/node_manager.cc (hunk 12 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: revised.

File: src/ray/raylet/node_manager.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: done. thanks!

File: src/ray/raylet/node_manager.cc (hunk 10 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: yes, that looks cleaner.

File: src/ray/raylet/node_manager.cc (hunk 13 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: Thanks for pointing this out. In this case we cannot guarantee the order because the `NotifyUnblocked` message and GRPC reply are sent via different connections.
To ensure the order, I modified the worker to send a `TaskDone` message when it finishes a task, and raylet will just treat this message in the same way as a `GetTask`.

**Comment by raulchen** [comment]
I just merged a PR. Can you resolve the conflicts?

**Comment by zhijunfu** [comment]
> I just merged a PR. Can you resolve the conflicts?

Updated. thanks.

File: src/ray/raylet/node_manager.cc [hunk 20]
```diff
@@ -765,6 +765,9 @@ void NodeManager::ProcessClientMessage(
   case protocol::MessageType::GetTask: {
     ProcessGetTaskMessage(client);
   } break;
+  case protocol::MessageType::TaskDone: {
+    ProcessGetTaskMessage(client);
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: check `UsePush == true` here? so is `GetTask`.

File: src/ray/raylet/scheduling_resources.h [hunk 21]
```diff
@@ -443,6 +443,11 @@ class ResourceIdSet {
   std::vector<flatbuffers::Offset<ray::protocol::ResourceIdSetInfo>> ToFlatbuf(
       flatbuffers::FlatBufferBuilder &fbb) const;
 
+  /// \brief Serialize this object as a string.
+  ///
+  /// \return A serialized string of this object.
+  const std::string Serialize() const;
```
Comment (raulchen) [inline src/ray/raylet/scheduling_resources.h]: Add a TODO explaining that this can be removed when raylet client is switched to grpc?

File: src/ray/raylet/node_manager.cc [hunk 22]
```diff
@@ -835,14 +838,20 @@ void NodeManager::ProcessClientMessage(
...
+                                         client, client_call_manager_);
   if (message->is_worker()) {
     // Register the new worker.
+    bool use_push_task = worker->UsePush();
+    auto connection = worker->Connection();
     worker_pool_.RegisterWorker(std::move(worker));
-    DispatchTasks(local_queues_.GetReadyTasksWithResources());
+    if (use_push_task) {
+      // only call `ProcessGetTaskMessage` when push mode is used.
+      ProcessGetTaskMessage(connection);
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc]: Nit, `ProcessGetTaskMessage`'s name is kind of improper here. 
I would rename it to something like `OnWorkerIdle` and pass the worker as the parameter.

**Review by raulchen — APPROVED** [review]
Looks good overall. Just left a few minor comments.

File: src/ray/raylet/node_manager.cc (hunk 20 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: ok

File: src/ray/raylet/node_manager.cc (hunk 22 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: I agree that the naming here is not best, but also I'm a bit hesitant to add another function which is exactly the same as `ProcessGetTaskMessage` with just a different name. One the other hand, we already call `ProcessDisconnectClientMessage` when we want to disconnect a worker, so it might also be ok here.

File: src/ray/core_worker/transport/raylet_transport.cc [hunk 23]
```diff
@@ -26,6 +28,12 @@ void CoreWorkerRayletTaskReceiver::HandleAssignTask(
   const raylet::Task task(request.task());
   const auto &spec = task.GetTaskSpecification();
   auto status = task_handler_(spec);
+  // Notify raylet the current task is done. This is to ensure that the task
+  // is marked as finished by raylet only after previous raylet client calls are
+  // completed. The rpc `done_callback` is sent via a different connection
+  // from raylet client connection, so it cannot guarantee the rpc reply arrives
+  // at raylet after a previous `NotifyUnblocked` message.
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/raylet_transport.cc]: I'm a little confused by this comment...the second sentence makes it sound like the raylet will not get this callback until after all other calls have completed, but then the third sentence seems to say the opposite? What exactly is the guarantee of `TaskDone`?

File: src/ray/raylet/node_manager.cc [hunk 24]
```diff
@@ -763,6 +763,11 @@ void NodeManager::ProcessClientMessage(
     ProcessRegisterClientRequestMessage(client, message_data);
   } break;
   case protocol::MessageType::GetTask: {
+    RAY_CHECK(!registered_worker->UsePush());
+    ProcessGetTaskMessage(client);
+  } break;
+  case protocol::MessageType::TaskDone: {
+    RAY_CHECK(registered_worker->UsePush());
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: I'm confused about the difference between this and `GetTask`? When do we call one and/or the other?

File: src/ray/raylet/node_manager.cc (hunk 22 above)
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc]: I think we can just rename `ProcessGetTaskMessage`. To match the rest of the code, something like `HandleWorkerAvailable` would be good (I think similar to what you had in a previous commit).

File: src/ray/raylet/node_manager.cc [hunk 25]
```diff
@@ -837,33 +842,17 @@ void NodeManager::ProcessRegisterClientRequestMessage(
...
-        return;
-      }
-
-      // Initialize a rpc client to the new worker.
-      std::unique_ptr<rpc::WorkerTaskClient> grpc_client(
-          new rpc::WorkerTaskClient("127.0.0.1", worker->Port(), client_call_manager_));
-      worker_clients_.emplace(client_id, std::move(grpc_client));
-    }
-
+    bool use_push_task = worker->UsePush();
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: Nit, please move this to be above the `if (use_push_task)` statement that uses it below, and move the `auto connection ...` line inside the if block.

**Review by stephanie-wang — APPROVED** [review]
Thanks, this is looking great! I left a couple comments that I would like addressed before we merge, but otherwise looks good.

File: src/ray/core_worker/transport/raylet_transport.cc (hunk 23 above)
Comment (zhijunfu) [reply src/ray/core_worker/transport/raylet_transport.cc]: Sorry for the confusion. The third sentence is actually saying this is not possible with rpc `done_callback`, and that's why we need to use `TaskDone`. I'll update the comments and make it clear.

File: src/ray/raylet/node_manager.cc (hunk 24 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: The rule is simple: 
- for legacy workers they still use `GetTask`, which indicates previous task is done and fetches a new task from raylet; 
- for workers based on c++ core worker, they will use `TaskDone`, just to indicate a task is finished, and then raylet will push a new task to the worker via grpc.

File: src/ray/raylet/node_manager.cc (hunk 25 above)
Comment (zhijunfu) [reply src/ray/raylet/node_manager.cc]: This is actually intentional as `worker_pool_.RegisterWorker` does a `std::move` and `worker` is no longer valid after that line.

**Comment by zhijunfu** [comment]
Thanks for reviewing! I think all comments are addressed.

**Comment by zhijunfu** [comment]
travis, retest this please

**Comment by stephanie-wang** [comment]
I couldn't trigger a travis build for some reason...
